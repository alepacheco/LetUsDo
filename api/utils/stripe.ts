import * as Stripe from 'stripe';

type HandleErrorArgs = {
    status?: string;
    next_action?: {
        type: string;
    };
    client_secret?: string;
};

export const handleError = async (func: () => Promise<HandleErrorArgs>) => {
    try {
        const { status, next_action, client_secret } = await func();

        if (status === 'requires_action' && next_action && next_action.type === 'use_stripe_sdk') {
            return {
                status: 200,
                response: {
                    requires_action: true,
                    payment_intent_client_secret: client_secret
                }
            };
        }
        if (status === 'succeeded') {
            return {
                status: 200,
                response: {
                    success: true
                }
            };
        }
    } catch (error) {
        return {
            status: 500,
            response: {
                error
            }
        };
    }

    return {
        status: 500,
        error: 'Unknown error while executing payment'
    };
};

type ExecutePaymentMethod = {
    payment_method_id: string;
    amount: number;
    taskText: string;
    metadata: {
        email: string;
        remoteAddress: string;
        name: string;
        phone: string;
    };

    stripe: Stripe;
};
export const executePaymentMethod = async ({
    payment_method_id,
    amount,
    taskText,
    metadata,
    stripe
}: ExecutePaymentMethod) => handleError(() => stripe.paymentIntents.create({
    payment_method: payment_method_id,
    amount,
    description: `Task details: ${taskText}`,
    currency: 'gbp',
    receipt_email: metadata.email,
    confirmation_method: 'manual',
    confirm: true,
    metadata
}));


type ExecutePaymentIntent = {
    payment_intent_id: string;
    stripe: Stripe;
};
export const executePaymentIntent = async ({ stripe, payment_intent_id }: ExecutePaymentIntent) =>
    handleError(() => stripe.paymentIntents.confirm(
        payment_intent_id
    ));
