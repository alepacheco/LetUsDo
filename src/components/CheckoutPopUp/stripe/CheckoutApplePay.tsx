import React from 'react';
import { injectStripe, PaymentRequestButtonElement } from 'react-stripe-elements';
import * as actions from 'src/actions/taskModalActions';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

class CheckoutApplePay extends React.Component<any, any> {
  constructor(props: any) {
    super(props);

    const paymentRequest = props.stripe.paymentRequest({
      country: 'GB',
      currency: 'gbp',
      total: {
        label: 'Task fixed price',
        amount: 99
      },
      requestPayerName: true,
      requestPayerEmail: true,
      requestPayerPhone: true
    });

    paymentRequest.on('token', ({ complete, token, ...data }: any) => {
      console.log('Received Stripe token: ', token);
      console.log('Received customer information: ', data);
      complete('success');
    });

    paymentRequest.canMakePayment().then((result: any) => {
      this.setState({ canMakePayment: !!result });
      this.props.actions.setApplePayAvailable(!!result);

      console.log({ canMakePayment: !!result });
    });

    this.state = {
      canMakePayment: false,
      paymentRequest
    };
  }

  render() {
    return this.state.canMakePayment ? (
      <PaymentRequestButtonElement
        paymentRequest={this.state.paymentRequest}
        className="PaymentRequestButton"
      />
    ) : null;
  }
}

const mapDispatchToProps = (dispatch: any) => ({
  actions: bindActionCreators(actions, dispatch)
});

export default connect(
  () => ({}),
  mapDispatchToProps
)(injectStripe(CheckoutApplePay));
