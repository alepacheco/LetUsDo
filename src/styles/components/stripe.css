.card-numer-field {
    border: 2px solid #dee2e6;
    border-radius: 2px;
    padding: 8px;
    margin: 8px 0;
}

.pay-button-wrapper {
    text-align: right;
}

.pay-button {

}