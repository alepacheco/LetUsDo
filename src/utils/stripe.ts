import axios from 'axios';
import { ReactStripeElements } from 'react-stripe-elements';

type TryPayment = {
  taskText: string;
  email?: string;
  name?: string;
  phone?: string;
  payment_method_id?: string;
  payment_intent_id?: string;
  stripe: ReactStripeElements.StripeProps;
};

export const tryPayment = async ({
  taskText,
  email,
  name,
  phone,
  payment_method_id,
  payment_intent_id,
  stripe
}: TryPayment): Promise<boolean | undefined> => {
  const {
    data: { requires_action, payment_intent_client_secret, success, error }
  } = await axios.post('/api/createPayment', {
    payment_method_id,
    payment_intent_id,
    taskText,
    metadata: {
      email,
      name,
      phone,
    }
  });

  if (error) {
    throw new Error(error);
  }

  if (success) {
    return true;
  }

  if (requires_action) {
    // @ts-ignore
    const { error, paymentIntent } = await stripe.handleCardAction(payment_intent_client_secret);

    if (error || !paymentIntent.id) {
      throw error;
    }

    return tryPayment({
      payment_intent_id: paymentIntent.id,
      taskText,
      email,
      name,
      phone,
      stripe
    });
  }
};

type SubmitProps = {
  email: string;
  stripe: ReactStripeElements.StripeProps;
  taskText: string;
};
export const submitPayment = async ({ email, stripe, taskText }: SubmitProps): Promise<boolean> => {
  try {
    const { paymentMethod, error } = await stripe.createPaymentMethod('card');

    if (error || !paymentMethod || !paymentMethod.id) {
      throw error;
    }

    return (
      (await tryPayment({
        taskText,
        email,
        payment_method_id: paymentMethod.id,
        stripe
      })) || false
    );
  } catch (error) {
    console.log(error);
    return false;
  }
};
