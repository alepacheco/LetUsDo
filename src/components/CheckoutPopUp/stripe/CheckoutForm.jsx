/* eslint-disable no-console */
import React from 'react';
import { bindActionCreators } from 'redux';
import Button from 'react-bootstrap/Button';
import Badge from 'react-bootstrap/Badge';
import { connect } from 'react-redux';
import { CardElement, injectStripe } from 'react-stripe-elements';
import '../../../styles/components/stripe.css';
import PropTypes from 'prop-types';
import axios from 'axios';
import * as actions from '../../../actions/taskModalActions';

const submit = async ({ email, stripe, taskText, setDialog }) => {
  const { token, error: createTokenError } = await stripe.createToken({
    name: 'Let Us Do'
  });
  if (!token) {
    console.log(createTokenError);
    return;
  }

  try {
    await axios.post('/api/createPayment', {
      token,
      email,
      taskText
    });

    setDialog('purchaseCompleted');
  } catch (error) {
    console.log(error.message);
  }
};

export const CheckoutForm = ({ email, stripe, taskText, actions: { setDialog }, validEmail }) => {
  const onClick = () =>
    submit({
      email,
      stripe,
      taskText,
      setDialog
    });

  return (
    <div className="checkout">
      <div className="card-numer-field">
        <CardElement />
      </div>
      <div className="pay-button-wrapper">
        <Button className="pay-button" onClick={onClick} disabled={!validEmail}>
          Pay{' '}
          <Badge pill variant="light">
            20Â£
          </Badge>
        </Button>
      </div>
    </div>
  );
};

const mapStateToProps = state => ({
  validEmail: state.checkoutPopUp.validEmail,
  email: state.checkoutPopUp.email,
  taskText: state.taskModal.taskText
});
const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(actions, dispatch)
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(injectStripe(CheckoutForm));

CheckoutForm.propTypes = {
  validEmail: PropTypes.bool.isRequired,
  email: PropTypes.string.isRequired,
  taskText: PropTypes.string.isRequired,
  stripe: PropTypes.object.isRequired,
  actions: PropTypes.shape({
    setDialog: PropTypes.func
  }).isRequired
};
