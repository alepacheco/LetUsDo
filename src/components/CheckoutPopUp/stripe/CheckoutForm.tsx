/* eslint-disable no-console */
import React from 'react';
import { bindActionCreators } from 'redux';
import Button from 'react-bootstrap/Button';
import Badge from 'react-bootstrap/Badge';
import { connect } from 'react-redux';
import { CardElement, injectStripe, ReactStripeElements } from 'react-stripe-elements';
import { trackEvent } from 'src/utils/analytics';
import 'src/styles/components/stripe.css';
import { submitPayment } from 'src/utils/stripe';
// @ts-ignore
import * as actions from 'src/actions/taskModalActions';

const tryApplePay = async ({ stripe }: { stripe: ReactStripeElements.StripeProps }) => {
  const paymentRequest = stripe.paymentRequest({
    country: 'GB', // TODO check
    currency: 'gbp',
    total: {
      label: 'Price',
      amount: 50
    },
    requestPayerName: true,
    requestPayerEmail: true
  });

  const canUsePayBotton = await paymentRequest.canMakePayment();
  if (canUsePayBotton) {
    // @ts-ignore
    const elements = stripe.elements();

    const prButton = elements.create('paymentRequestButton', {
      paymentRequest
    });

    prButton.mount('#checkout-bottom');

    const element = document.getElementById('checkout-form');
    if (element) {
      element.style.display = 'none';
    }
  } else {
    const element = document.getElementById('checkout-bottom');
    if (element) {
      element.style.display = 'none';
    }
    return;
  }

  // @ts-ignore
  const elements = stripe.elements();

  const prButton = elements.create('paymentRequestButton', {
    paymentRequest
  });

  prButton.mount('#checkout-bottom');

  const element = document.getElementById('checkout-form');
  if (element) {
    element.style.display = 'none';
  }

  paymentRequest.on('paymentmethod', async ev => {
    // @ts-ignore
    const { error: confirmError, paymentIntent } = await stripe.confirmPaymentIntent(
      process.env.STRIPE_FRONT,
      {
        payment_method: ev.paymentMethod.id
      }
    );

    if (confirmError || !process.env.STRIPE_FRONT) {
      // Report to the browser that the payment failed, prompting it to
      // re-show the payment interface, or show an error message and close
      // the payment interface.
      console.log('apple pay failed');
      ev.complete('fail');
    } else {
      // Report to the browser that the confirmation was successful, prompting
      // it to close the browser payment method collection interface.
      console.log('apple pay collected');
      ev.complete('success');
      // Let Stripe.js handle the rest of the payment flow.
      const { error } = await stripe.handleCardPayment(process.env.STRIPE_FRONT);
      if (error) {
        console.log('apple pay failed');
        // The payment failed -- ask your customer for a new payment method.
      } else {
        console.log('apple pay payed');
        // The payment has succeeded.
      }
    }
  });
};

export const CheckoutForm: React.FC<{
  email: string;
  taskText: string;
  stripe?: ReactStripeElements.StripeProps;
  validEmail: boolean;
  actions: { setDialog: (state: string) => void };
}> = ({ email, stripe, taskText, actions: { setDialog }, validEmail }) => {
  if (stripe) {
    // tryApplePay({ stripe });
  }

  const onClick = async () => {
    console.log('Set loading here but keep the element alive.');
    trackEvent({
      category: 'click',
      action: 'pay button clicked'
    });
    trackEvent({
      category: 'purchase',
      action: 'purchase initiated'
    });

    if (!stripe) {
      setDialog('purchaseFailed');

      return;
    }

    const completed = await submitPayment({
      email,
      stripe,
      taskText
    });
    if (completed) {
      setDialog('purchaseCompleted');
      trackEvent({
        category: 'purchase',
        action: 'purchase completed'
      });
    } else {
      setDialog('purchaseFailed');
      trackEvent({
        category: 'purchase',
        action: 'purchase failed'
      });
    }
  };

  return (
    <div className="checkout">
      <div id="checkout-bottom" />
      <div id="checkout-form">
        <div className="card-numer-field">
          <CardElement />
        </div>
        <div className="pay-button-wrapper">
          <Button className="pay-button" onClick={onClick} disabled={!validEmail}>
            Pay{' '}
            <Badge pill variant="light">
              20Â£
            </Badge>
          </Button>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = (state: any) => ({
  validEmail: state.checkoutPopUp.validEmail,
  email: state.checkoutPopUp.email,
  taskText: state.taskModal.taskText
});
const mapDispatchToProps = (dispatch: any) => ({
  actions: bindActionCreators(actions, dispatch)
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(injectStripe(CheckoutForm));
