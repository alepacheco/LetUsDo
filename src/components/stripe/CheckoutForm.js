import React from 'react';
import Button from 'react-bootstrap/Button';
import Badge from 'react-bootstrap/Badge';
import { connect } from 'react-redux';
import { CardElement, injectStripe } from 'react-stripe-elements';
import '../../styles/components/stripe.css';
import PropTypes from 'prop-types';

class CheckoutForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = { complete: false };
    this.submit = this.submit.bind(this);
  }

  async submit() {
    const { token } = await this.props.stripe.createToken({ name: "We Do" });
    
    if (!token) {
      console.log("No card!");
      return;
    }

    const response = await fetch("/api/createPayment", {
      method: "POST",
      headers: { "Content-Type": "text/plain" },
      body: JSON.stringify(token)
    });

    if (response.ok) console.log("Purchase Complete!")
    if (!response.ok) console.log("Error!")
    if (response.ok) this.setState({ complete: true });

  }

  render() {
    if (this.state.complete) return <h1>Purchase Complete</h1>;

    return (
      <div className="checkout">
        <div className="card-numer-field">
          <CardElement />
        </div>
        <div className="pay-button-wrapper">
          <Button
            className="pay-button"
            onClick={this.submit}
            disabled={!this.props.validateEmail}>
              Pay <Badge pill variant="light"> 20Â£</Badge>
          </Button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  validateEmail: state.checkoutPopUp.validateEmail
});

export default connect(
  mapStateToProps
)(injectStripe(CheckoutForm));

CheckoutForm.propTypes = {
  validateEmail: PropTypes.string
};