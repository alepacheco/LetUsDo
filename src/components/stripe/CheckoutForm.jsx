/* eslint-disable no-console */
import React from 'react';
import Button from 'react-bootstrap/Button';
import Badge from 'react-bootstrap/Badge';
import { connect } from 'react-redux';
import { CardElement, injectStripe } from 'react-stripe-elements';
import '../../styles/components/stripe.css';
import PropTypes from 'prop-types';
import axios from 'axios';

export class CheckoutForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = { complete: false };
    this.submit = this.submit.bind(this);
    console.log(process.env.STRIPE_FRONT);

    // eslint-disable-next-line no-undef
    console.log(STRIPE_FRONT);
  }

  async submit() {
    const { token, error: createTokenError } = await this.props.stripe.createToken({
      name: 'Let Us Do'
    });
    if (!token) {
      console.log(createTokenError);
      return;
    }

    try {
      const response = await axios.post('/api/createPayment', {
        token,
        email: this.props.email,
        taskText: this.props.taskText
      });

      console.log({ response });
      this.setState({ complete: true });
    } catch (error) {
      console.log(error.message);
    }
  }

  render() {
    // TODO improve
    if (this.state.complete) {
      return <h1>Purchase Complete, we will get in touch with updates to your task</h1>;
    }

    return (
      <div className="checkout">
        <div className="card-numer-field">
          <CardElement />
        </div>
        <div className="pay-button-wrapper">
          <Button className="pay-button" onClick={this.submit} disabled={!this.props.validEmail}>
            Pay{' '}
            <Badge pill variant="light">
              20Â£
            </Badge>
          </Button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  validEmail: state.checkoutPopUp.validEmail,
  email: state.checkoutPopUp.email,
  taskText: state.taskModal.taskText
});

export default connect(mapStateToProps)(injectStripe(CheckoutForm));

CheckoutForm.propTypes = {
  validEmail: PropTypes.bool,
  email: PropTypes.string,
  taskText: PropTypes.string,
  stripe: PropTypes.object
};
