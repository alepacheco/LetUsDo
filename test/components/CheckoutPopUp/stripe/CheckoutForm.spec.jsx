/* eslint-disable no-undef */
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { CheckoutForm } from '../../../../src/components/CheckoutPopUp/stripe/CheckoutForm';

describe('<CheckoutForm />', () => {
  it('renders', () => {
    const wrapper = shallow(
      <CheckoutForm
        actions={{}}
        stripe={{}}
        validEmail
        email="email@example.com"
        taskText="this is a test"
      />
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
