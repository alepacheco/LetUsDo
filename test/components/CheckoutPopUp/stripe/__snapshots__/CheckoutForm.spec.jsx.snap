// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckoutForm /> renders 1`] = `
<div
  className="checkout"
>
  <div
    className="card-numer-field"
  >
    <CardElement
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onReady={[Function]}
    />
  </div>
  <div
    className="pay-button-wrapper"
  >
    <Button
      active={false}
      className="pay-button"
      disabled={false}
      onClick={[Function]}
      type="button"
      variant="primary"
    >
      Pay
       
      <Badge
        pill={true}
        variant="light"
      >
        20Â£
      </Badge>
    </Button>
  </div>
</div>
`;
